GSE198256_ANALYSIS_modified.R---
title: "DAY2 GSE198256"
author: "David Gomez-Cabrero"
date: "2024-January"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Experimental design

Lets review experimental design from a practical perspective

```{r read data}

# Read data
urld <- "https://www.ncbi.nlm.nih.gov/geo/download/?format=file&type=rnaseq_counts"
path <- paste(urld, "acc=GSE198256", "file=GSE198256_raw_counts_GRCh38.p13_NCBI.tsv.gz", sep="&");
GSE198256_count <- as.matrix(data.table::fread(path, header=T, colClasses="integer"), rownames=1)

# Read Meta data
library(GEOquery)
gds <- getGEO("GSE198256")
Meta_GSE198256 <- pData(gds$GSE198256_series_matrix.txt.gz@phenoData)
Group <- Meta_GSE198256[,c("disease state:ch1")]

dim(GSE198256_count)
Group

```

## Limma: Normalize and set design

```{r Normalize and set design}

# set DGE class
require(limma)
require(edgeR)
dge <- DGEList(counts=GSE198256_count)

# Make sure on the metadata
rownames(Meta_GSE198256)==colnames(GSE198256_count)
Group[Group=="Covid19: Acute infection"] <- "Covid19AI"
Group[Group=="Covid19: Recovery 3Mo"] <- "Covid193Mo"
Group[Group=="Covid19: Recovery 6Mo"] <- "Covid196Mo"
design <- model.matrix(~ Group )

# Filter
keep <- filterByExpr(dge, design=design)
dge <- dge[keep,,keep.lib.sizes=FALSE]

# Normalization
dge <- calcNormFactors(dge)


```

## Limma: Voom or Trend?

```{r Voom or Trend}

## Trend

# If the sequencing depth is reasonably consistent across the RNA samples, then the simplest and most robust approach to differential exis to use limma-trend. This approach will usually work well if the ratio of the largest library size to the smallest is not more than about 3-fold.
logCPM <- cpm(dge, log=TRUE, prior.count=3)
 # The prior count is used here to damp down the variances of logarithms of low counts.
fit <- lmFit(logCPM, design)

fit <- eBayes(fit, trend=TRUE)
 # logical, should an intensity-dependent trend be allowed for the prior variance? If FALSE then the prior variance is constant. Alternatively, trend can be a row-wise numeric vector, which will be used as the covariate for the prior variance.
 # The use of eBayes or treat with trend=TRUE is known as the limma-trend method (Law et al, 2014; Phipson et al, 2016). With this option, an intensity-dependent trend is fitted to the prior variances s2.prior
trend_all <- topTable(fit, coef=ncol(design), number=Inf)
trend_all

## Voom

# When the library sizes are quite variable between samples, then the voom approach is theoretically more powerful than limma-trend.
v <- voom(dge, design, plot=TRUE)
# The voom method is similar in purpose to the limma-trend method, which uses eBayes or treat with trend=TRUE. The voom method incorporates the mean-variance trend into the precision weights, whereas limma-trend incorporates the trend into the empirical Bayes moderation. The voom method takes into account the sequencing depths (library sizes) of the individual columns of counts and applies the mean-variance trend on an individual observation basis. limma-trend, on the other hand, assumes that the library sizes are not wildly different and applies the mean-variance trend on a genewise basis. As noted by Law et al (2014), voom should be more powerful than limma-trend if the library sizes are very different but, otherwise, the two methods should give similar results.
fit <- lmFit(v, design)
fit <- eBayes(fit)
voom_all <- topTable(fit, coef=ncol(design), number=Inf)
voom_all
```

## ACTIVITY 1:

-   How would you compare the results between voom and trend?

    Two volcano plots were generated to compare the data. Both plots show a distribution of log-fold changes along the x-axis. The spread and distribution give an indication of how differentially expressed genes are between the conditions being tested. A wider spread suggests a greater difference in expression. The significant genes are colored in red. We can visually assess whether the same genes are showing up as significant in both analyses. Moreover, the data was further analysed to learn that the trend and voom analysis share 1168 significantly DEGs and disagree in 431. The significant overlap of 1168 significantly DEGs suggests that there is a core set of genes that both methods agree upon despite potential differences in normalization techniques and variance estimation. The discrepancy in 431 genes might be due to how each method handles the data, particularly with respect to the normalization of library sizes and the statistical models applied.

-   Is it required to run more than analysis?

    Running multiple analyses like voom and trend can be beneficial, especially in scenarios with varied sequencing depths across samples. Since voom accounts for mean-variance trends at an individual observation level and is generally more suitable for datasets with high variability in library sizes, while limma-trend applies a genewise approach assuming not too extreme variability, comparing both can provide a comprehensive view of the data. There are some other analysis we can perform to compare the results we obtain from both models, specially when we have unequal library sizes in our experiments. Some of these analysis include:

    -   Calculating the **correlation** between the log-fold changes from both analyses to assess their overall similarity.

    -   Getting a **Bland-Altman Plot**, which can help assess agreement between two quantitative measurements plotting the average of two measurements against their difference.

    -   Getting an **MA plot**, a popular method for visualizing the differences between measurements from high-throughput genomics experiments. It plots the average intensity (A) versus the log ratio (M) of the counts.

-   What exactly are we asking with this differential expression?

    With differential expression analysis, the primary goal is to determine which genes are upregulated or downregulated between different conditions or states. The key questions often asked include:

    -   **Biological Significance** -\> Which genes change their expression levels in response to a certain condition or treatment, and are these changes biologically meaningful?

    -   **Molecular Mechanisms** -\> Can the patterns of gene expression reveal underlying molecular mechanisms that may explain a biological process or the pathogenesis of a disease?

    -   **Biomarker Discovery** -\> Are there potential biomarkers that can be identified for diagnostic, prognostic, or therapeutic purposes?

    -   **Gene Function and Interaction**: How do differentially expressed genes interact with each other, and what can this tell us about functional pathways and networks?

    By addressing these questions, we aim to uncover the molecular patterns associated with the conditions of interest, which can lead to further hypothesis generation and experimental studies. Differential expression analysis is often the starting point for deeper functional genomics investigations and can provide insights into the complexity of gene regulation.

```{r ACTIVITY 1}
# Assuming trend_all and voom_all are your dataframes
combined_df <- merge(trend_all, voom_all, by="row.names", suffixes=c("_trend", "_voom"))

# For Trend
plot(combined_df$logFC_trend, -log10(combined_df$P.Value_trend),
     xlab="LogFC", ylab="-log10 P-Value",
     col=ifelse(combined_df$adj.P.Val_trend < signif_threshold, "red", "black"),
     main="Volcano Plot - Trend")

# For Voom
plot(combined_df$logFC_voom, -log10(combined_df$P.Value_voom),
     xlab="LogFC", ylab="-log10 P-Value",
     col=ifelse(combined_df$adj.P.Val_voom < signif_threshold, "red", "black"),
     main="Volcano Plot - Voom")

signif_threshold <- 0.05 #threshold for adjusted p value
sig_genes_trend <- row.names(combined_df)[combined_df$adj.P.Val_trend < signif_threshold]
sig_genes_voom <- row.names(combined_df)[combined_df$adj.P.Val_voom < signif_threshold]

# Find common significant genes
common_sig_genes <- intersect(sig_genes_trend, sig_genes_voom)
non_common_sig_genes <- setdiff(sig_genes_trend, sig_genes_voom)
length(common_sig_genes)
length(non_common_sig_genes)

```

## ACTIVITY 2:

-   Plan the next analysis: questions, steps,...\
    Some questions to adress: What are the biological implications of the DEGs identified in each contrast?Which biological pathways are enriched for the DEGs in each contrast? What are the biological processes, cellular components, and molecular functions associated with the DEGs? How do the DEGs interact at the protein level, and what are the key regulatory genes? How does gene expression evolve from acute infection to recovery at 6 months? Are there potential biomarkers or therapeutic targets among the DEGs for COVID-19 infection and recovery?\
    \
    We will expand our contrasts to explore more nuanced relationships within our data. Utilizing the voom transformation from the limma package, we will refine our linear models to encompass additional comparisons, such as between "Covid19 at 6 months" vs. "Healthy" and "Covid19 Acute Infection" vs. "Healthy". By constructing a contrast matrix and applying it to our model, we aim to extract differential expression results for these specific pairwise comparisons. We'll extract comprehensive tables for all contrasts using `topTable`, resulting tables will be stored for in-depth evaluation, setting the stage for subsequent GSEA to uncover the underlying biological pathways affected by these conditions.

```{r ACTIVITY 2}

# PACKAGES
#BiocManager::install("clusterProfiler", update = FALSE)
#BiocManager::install("ggupset", update = FALSE)
#BiocManager::install("msigdbr", update = FALSE)
#BiocManager::install("org.Hs.eg.db", update = FALSE)

library(clusterProfiler)
library(msigdbr)
library(org.Hs.eg.db)
library(magrittr)

# Add more contrasts

v <- voom(dge, design, plot=TRUE)
colnames(design) <- c("Intercept","Covid196Mo","Covid19AI","Healthy")
fit <- lmFit(v, design)
contrast.matrix <- makeContrasts(Covid19AI-Healthy, Healthy, 
                                 Covid196Mo-Healthy,    
                                 levels=design)
fit2 <- contrasts.fit(fit, contrast.matrix)
fit2 <- eBayes(fit2)
all_contrasts <- topTable(fit2, number=Inf) 
Covid19AI_Healthy_contrast <- topTable(fit2,coef=1, number=Inf) 
Healthy_contrast <- topTable(fit2,coef=2, number=Inf) 
Covid196Mo_Healthy_contrast <- topTable(fit2,coef=3, number=Inf) 
# topTable(fit2,coef=4) try this! -> doesn't work because subscript is out of bounds

# Store all of them
all_contrasts
Covid19AI_Healthy_contrast
Healthy_contrast
Covid196Mo_Healthy_contrast

```

## ORA and Gene Set Enrichment analysis.

-   What do we need to do the analysis?\
    For Over-Representation Analysis (ORA) and Gene Set Enrichment Analysis (GSEA), the primary requirement is a list of genes that have been identified as differentially expressed (DEGs) between conditions of interest, usually along with associated statistics such as p-values or log fold changes. These lists are then tested for enrichment in predefined sets of genes that are associated with particular biological functions, pathways, or other attributes.\
-   What are the tools required?\
    To perform ORA and GSEA, you would need:
    1.  **Gene List**: A ranked list of genes based on their differential expression statistics. For ORA, we would typically use a list of genes that pass a certain significance threshold, while for GSEA, the full list of genes ranked by their expression changes or associated statistics is used.

    2.  **Gene Sets**: Collections of genes grouped based on shared biological functions, pathways, or ontologies. These can be sourced from databases like MSigDB, KEGG, GO, Reactome, etc.

    3.  **Mapping Data**: Correspondence between different gene identifiers (e.g., mapping between ENSEMBL IDs and ENTREZ IDs) to ensure that the gene lists and gene sets are compatible.

    4.  **Statistical Tools/Software**: Tools like **`clusterProfiler`**, **`limma`**, **`fgsea`**, or other R packages that can perform the statistical calculations to determine enrichment. These tools often include methods for correcting for multiple hypothesis testing, such as the Benjamini-Hochberg procedure.

    5.  **Annotation Packages**: Databases such as **`org.Hs.eg.db`** in Bioconductor, which provide mappings between gene identifiers, as well as additional gene information that may be required for the analysis.\

```{r Prepare ORA and GSEA}

keytypes(org.Hs.eg.db)

# If we want to shift annotations:
ENSEMBL_vector <- mapIds(
  # Replace with annotation package for the organism relevant to your data
  org.Hs.eg.db,
  # The vector of gene identifiers we want to map
  keys = rownames(GSE198256_count),
  # Replace with the type of gene identifiers in your data
  keytype = "ENTREZID",
  # Replace with the type of gene identifiers you would like to map to
  column = "ENSEMBL",
  # In the case of 1:many mappings, return the
  # first one. This is default behavior!
  multiVals = "first"
)

# We would like a data frame we can join to the differential expression stats
gene_key_df <- data.frame(
  ensembl_id = ENSEMBL_vector,
  entrez_id = names(ENSEMBL_vector),
  stringsAsFactors = FALSE
) %>%
  # If an Ensembl gene identifier doesn't map to a gene symbol, drop that
  # from the data frame
  dplyr::filter(!is.na(ensembl_id))


```

## Lets conduct ORA.

-   What do we need to do the analysis?\
    To conduct Over-Representation Analysis (ORA), you need a list of genes of interest, often those that are differentially expressed at a statistically significant level. You also require a background gene set, which represents the universe of all possible genes that could be differentially expressed in your experiment. Additionally, you need a collection of gene sets that represent biological pathways or processes for which you'll test enrichment.

-   What are the tools required?

    The tools we need for ORA include functions from the `clusterProfiler` and `enrichplot` R packages, as well as gene set data from the `msigdbr` package. Specifically:

    -   `topTable` from the `limma` package to obtain a table of differentially expressed genes at a specified p-value threshold and selected the gene identifiers as your list of genes of interest.
    -   Define the background set as the unique genes from your logCPM object.
    -   Retrieve human gene sets from the Molecular Signatures Database (MSigDB) using `msigdbr`, focusing on the **KEGG** pathways.
    -   The `enricher` function from `clusterProfiler` is employed to conduct ORA, using the genes of interest, the background set, and the KEGG pathways gene sets.
    -   Visualize results with `dotplot` and `upsetplot` from the `enrichplot` package.

-   Besides **KEGG**, there are several alternative pathway databases and gene set collections that we can consider for ORA, such as:

    -   **GO (Gene Ontology)**: Enrichment analysis using GO terms, which are divided into biological processes, cellular components, and molecular functions.

    -   **Reactome**: A curated database of pathways and reactions in human biology.

    -   **Biocarta**: A collection of pathway maps.

    -   **Hallmark gene sets**: Gene sets that represent specific well-defined biological states or processes and display coherent expression.

    -   **Custom gene sets**: We might also consider creating custom gene sets based on literature research or experimental evidence specific to your study.

```{r Run ORA}

# Step 1: determine genes of interest.
diff_table <- topTable(fit2,coef=1,p.value=0.01,number=10000) 
genes_dif<- rownames(diff_table )

# Step 2: determine background.

background_set <- unique(rownames(logCPM))

# Step 3: Determine gene sets.

msigdbr_species()
hs_msigdb_df <- msigdbr(species = "Homo sapiens")
head(hs_msigdb_df)

hs_kegg_df <- hs_msigdb_df %>%
  dplyr::filter(
    gs_cat == "C2", # This is to filter only to the C2 curated gene sets
    gs_subcat == "CP:KEGG" # This is because we only want KEGG pathways
  )

# Step 4: conduct ORA.

kegg_ora_results <- enricher(
  gene = genes_dif, # A vector of your genes of interest
  pvalueCutoff = 0.1, # Can choose a FDR cutoff
  pAdjustMethod = "BH", # Method to be used for multiple testing correction
  universe = background_set, # A vector containing your background set genes
  # The pathway information should be a data frame with a term name or
  # identifier and the gene identifiers
  TERM2GENE = dplyr::select(
    hs_kegg_df,
    gs_name,
    human_entrez_gene
  )
)


# Step 5: Visualize / explore

enrich_plot <- enrichplot::dotplot(kegg_ora_results)
enrich_plot

upset_plot <- enrichplot::upsetplot(kegg_ora_results)
upset_plot

# Step 6: EXERCISE: alternatives to KEGG?


# note: see mor examples: https://alexslemonade.github.io/refinebio-examples/03-rnaseq/pathway-analysis_rnaseq_01_ora.html#46_Over-representation_Analysis_(ORA) 

```

## Lets conduct GSEA.

```{r run GSEA}


# Step 1: determine genes of interest.
diff_table_all <- topTable(fit2,coef=1,p.value=1,number=nrow(logCPM)) 

# Step 2: determine background.

# Step 3: Determine gene sets.

msigdbr_species()
hs_msigdb_df <- msigdbr(species = "Homo sapiens")
head(hs_msigdb_df)

hs_kegg_df <- hs_msigdb_df %>%
  dplyr::filter(
    gs_cat == "C2", # This is to filter only to the C2 curated gene sets
    gs_subcat == "CP:KEGG" # This is because we only want KEGG pathways
  )

# Step 4: conduct GSEA

list_ordered <- diff_table_all[,"B"]
names(list_ordered) <- rownames(diff_table_all)
  
set.seed(123)

gsea_results <- GSEA(
  geneList = list_ordered, # Ordered ranked gene list
  minGSSize = 25, # Minimum gene set size
  maxGSSize = 500, # Maximum gene set set
  pvalueCutoff = 0.05, # p-value cutoff
  eps = 0, # Boundary for calculating the p value #seed = 123, # Set seed to make results reproducible
  pAdjustMethod = "BH", # Benjamini-Hochberg correction
  TERM2GENE = dplyr::select(
    hs_kegg_df,
    gs_name,
    human_entrez_gene
  )
)

# Step 5: Visualize / explore
head(gsea_results@result)

gsea_result_df <- data.frame(gsea_results@result)
gsea_result_df %>%
  # This returns the 3 rows with the largest NES values
  dplyr::slice_max(NES, n = 3)

most_positive_nes_plot <- enrichplot::gseaplot(
  gsea_results,
  geneSetID = "KEGG_JAK_STAT_SIGNALING_PATHWAY",
  title = "KEGG_JAK_STAT_SIGNALING_PATHWAY",
  color.line = "#0d76ff"
)
most_positive_nes_plot

gsea_result_df %>%
  # Return the 3 rows with the smallest (most negative) NES values
  dplyr::slice_min(NES, n = 3)

most_negative_nes_plot <- enrichplot::gseaplot(
  gsea_results,
  geneSetID = "KEGG_SPLICEOSOME",
  title = "KEGG_SPLICEOSOME",
  color.line = "#0d76ff"
)
most_negative_nes_plot



# Step 6: EXERCISE: alternatives to KEGG?

# Step 7: EXERCISE: compare GSEA vs ORA?


```

## Lets conduct GeneSetCluster.

```{r prepare GSEA}
# Healthy vs Group Covid19 
# We prepare a function from the previous analysis

# Healthy vs Covid19AI 
Diff_HvsAI <- topTable(fit2,coef=1,p.value=1,number=nrow(logCPM))
# Healthy vs Covid196Mo 
Diff_Hvs6Mo <- topTable(fit2,coef=3,p.value=1,number=nrow(logCPM))


hs_msigdb_df <- msigdbr(species = "Homo sapiens")
hs_kegg_df <- hs_msigdb_df %>%
  dplyr::filter(
    gs_cat == "C2", # This is to filter only to the C2 curated gene sets
    gs_subcat == "CP:KEGG" # This is because we only want KEGG pathways
  )


doGSEA <- function(diff_table) {
  list_ordered <- diff_table[,"B"]
  names(list_ordered) <- rownames(diff_table)
  
  return(GSEA(
    geneList = list_ordered, # Ordered ranked gene list
    minGSSize = 25, # Minimum gene set size
    maxGSSize = 500, # Maximum gene set set
    pvalueCutoff = 0.05, # p-value cutoff
    eps = 0, # Boundary for calculating the p value
    seed = TRUE, # Set seed to make results reproducible
    pAdjustMethod = "BH", # Benjamini-Hochberg correction
    TERM2GENE = dplyr::select(
      hs_kegg_df,
      gs_name,
      human_entrez_gene
    )
  ))
}

GSEA_HvsAI <- doGSEA(Diff_HvsAI)
GSEA_Hvs6Mo <- doGSEA(Diff_Hvs6Mo)

path <- "/Users/velazqal/Documents/00_Bioinformatic pipelines"

write.csv(GSEA_HvsAI, file = paste0(path, "/GSEA_HvsAI.csv"), row.names = FALSE)
write.csv(GSEA_Hvs6Mo, file = paste0(path, "/GSEA_Hvs6Mo.csv"), row.names = FALSE)
```

```{r install GeneSetCluster}
######## Check if the required packages are installed ########
packages <- c("limma","stats","methods","RColorBrewer","clusterProfiler","GGally",
              "network","clustree","readxl","org.Hs.eg.db",
              "org.Mm.eg.db","cluster","factoextra","STRINGdb","WebGestaltR","stringr",
              "AnnotationDbi","ComplexHeatmap","GO.db","GetoptLong","bigstatsr","colorRamp2",
              "cowplot","doParallel","dplyr","foreach","ggdendro","ggnewscale","ggplot2",
              "ggtree","ggwordcloud","grid","httr","jsonlite","parallel","patchwork","pbapply",
              "reshape2","rgl","seriation","simplifyEnrichment","slam","tidyverse","umap",
               "utils","grDevices")

new.packages <- packages[!(packages %in% installed.packages()[,"Package"])]

install.packages(new.packages)

path <- "/Users/velazqal/Documents/00_Bioinformatic pipelines"
install.packages(path, repos=NULL, type='source')

```

```{r run GeneSetCluster}
library(GeneSetCluster)
GSEA.files <- paste0(path, "/", list.files(path, pattern = ".csv"))

# Load the data and create Pathway object
# Automatically for GSEA, GREAT or IPA
GSEA.Object1 <- LoadGeneSets(file_location = GSEA.files, 
                              groupnames= c("GSEA_Hvs6Mo", "GSEA_HvsAI"), # names of the groups
                              P.cutoff = 0.05, # cut off the p.adjust
                              Mol.cutoff = 15, # minimum number of genes per pathway
                              Source = "GSEA", # the analysis (GSEA, GREAT or IPA)
                              structure = "ENTREZID", # Gene type (SYMBOL, ENTREZID, ENSEMBLID)
                              Organism = "org.Hs.eg.db", # database: Homo Sapiens or Mus musculus
                              seperator = "/") # the separator used for listing genes

# IMPORTANT when created manually, it is assumed that the pathways have been filtered by p-value and minimum number of genes per pathway
# Make sure you have filtered your data
GSEA.Object1Manual <- ObjectCreator(Pathways = c(GSEA_HvsAI@result$ID, 
                                                 GSEA_Hvs6Mo@result$ID),
                                    Molecules = c(GSEA_HvsAI@result$core_enrichment, 
                                                  GSEA_Hvs6Mo@result$core_enrichment),
                                    Groups = c(rep("GSEA_HvsAI", times=nrow(GSEA_HvsAI@result)), 
                                               rep("GSEA_Hvs6Mo", times=nrow(GSEA_Hvs6Mo@result))),
                                    Pvalues = c(GSEA_HvsAI@result$p.adjust,  # optional
                                                GSEA_Hvs6Mo@result$p.adjust),
                                    enrichmentScore = c(GSEA_HvsAI@result$NES, # optional
                                                        GSEA_Hvs6Mo@result$NES), 
                                    structure = "ENTREZID", Type = "", sep = "/",
                                    Source = "GSEA", organism = "org.Hs.eg.db")

GSEA.Object2 <- CombineGeneSets(Object = GSEA.Object1,
                                combineMethod = "Standard", threads = 8)


OptimalGeneSets(Object = GSEA.Object2, 
                uniquePathway = FALSE, # consider all the pathways (also repeated) or the unique pathways
                method = "silhouette", max_cluster= 24, cluster_method = "kmeans", main= "Kmeans for 24 clusters")

OptimalGeneSets(Object = GSEA.Object2, 
                uniquePathway = TRUE, # consider all the pathways (also repeated) or the unique pathways
                method = "silhouette", max_cluster= 24, cluster_method = "kmeans", main= "Kmeans for 24 clusters")
# in both cases the optimal cluster is 2

GSEA.Object3 <- ClusterGeneSets(Object = GSEA.Object2, 
                                clusters = 2, # consider all the pathways (also repeated) or the unique pathways
                                method = "Hierarchical", # Hierarchical clustering or kmeans
                                order = "cluster",
                                molecular.signature = "All")

# plot results for both all pathways and unique pathways
plotnounique <- PlotGeneSets(GSEA.Object3, 
                             uniquePathways = FALSE, 
                             wordcloud = FALSE, # wordcloud only supported for GO terms
                             doORA = T) # do ora per cluster

plotunique <- PlotGeneSets(GSEA.Object3, 
                           uniquePathways = TRUE, 
                           wordcloud = FALSE, # wordcloud only supported for GO terms
                           doORA = T) # do ora per cluster


# let's say we are interested in exploring cluster 2 in plotunique. Lets break up this cluste for further analysis 

plotoptimalcluster2 <- OptimalGeneSets(Object = GSEA.Object3, 
                uniquePathway = TRUE, # consider all the pathways (also repeated) or the unique pathways
                cluster = 2, # which cluster
                method = "silhouette", max_cluster= 24, cluster_method = "kmeans", main= "Kmeans for 24 clusters in cluster 1")

plotoptimalcluster2 # optimal 2 break up cluster 2 in 2 clusters

GSEA.Object3breakup <- BreakUpCluster(GSEA.Object3, 
                                      breakup.cluster = 2, # which cluster
                                      sub.cluster = 2, # in how many cluster split up
                                      uniquePathways = TRUE) # conside unique pathways

plotuniquebreakup <- PlotGeneSets(GSEA.Object3breakup, 
                                  uniquePathways = TRUE, 
                                  wordcloud = FALSE, # wordcloud only supported for GO terms
                                  doORA = T) # do ora per cluster

plotuniquebreakup

# Now break up the cluster 1 
plotoptimalcluster1 <- OptimalGeneSets(Object = GSEA.Object3, 
                uniquePathway = TRUE, # consider all the pathways (also repeated) or the unique pathways
                cluster = 1, # which cluster
                method = "silhouette", max_cluster= 24, cluster_method = "kmeans", main= "Kmeans for 24 clusters in cluster 1")

plotoptimalcluster1 # optimal 1 break up cluster 1 in 9 clusters

GSEA.Object3breakup2 <- BreakUpCluster(GSEA.Object3breakup, 
                                      breakup.cluster = 1, # which cluster
                                      sub.cluster = 9, # in how many cluster split up
                                      uniquePathways = TRUE) # conside unique pathways

plotuniquebreakup2 <- PlotGeneSets(GSEA.Object3breakup2, 
                                   uniquePathways = TRUE, 
                                   wordcloud = FALSE, # wordcloud only supported for GO terms
                                   doORA = T) # do ora per cluster

plotuniquebreakup2


```

```{r explore results}
# plot results for both all pathways and unique pathways
plotnounique <- PlotGeneSets(GSEA.Object3, 
                             uniquePathways = FALSE, 
                             wordcloud = FALSE, # wordcloud only supported for GO terms
                             doORA = T) # do ora per cluster

plotunique <- PlotGeneSets(GSEA.Object3, 
                           uniquePathways = TRUE, 
                           wordcloud = FALSE, # wordcloud only supported for GO terms
                           doORA = T) # do ora per cluster


# let's say we are interested in exploring cluster 2 in plotunique. Lets break up this cluste for further analysis 

plotoptimalcluster2 <- OptimalGeneSets(Object = GSEA.Object3, 
                uniquePathway = TRUE, # consider all the pathways (also repeated) or the unique pathways
                cluster = 2, # which cluster
                method = "silhouette", max_cluster= 24, cluster_method = "kmeans", main= "Kmeans for 24 clusters in cluster 1")

plotoptimalcluster2 # optimal 2 break up cluster 2 in 2 clusters

GSEA.Object3breakup <- BreakUpCluster(GSEA.Object3, 
                                      breakup.cluster = 2, # which cluster
                                      sub.cluster = 2, # in how many cluster split up
                                      uniquePathways = TRUE) # conside unique pathways

plotuniquebreakup <- PlotGeneSets(GSEA.Object3breakup, 
                                  uniquePathways = TRUE, 
                                  wordcloud = FALSE, # wordcloud only supported for GO terms
                                  doORA = T) # do ora per cluster

plotuniquebreakup

# Now break up the cluster 1 
plotoptimalcluster1 <- OptimalGeneSets(Object = GSEA.Object3, 
                uniquePathway = TRUE, # consider all the pathways (also repeated) or the unique pathways
                cluster = 1, # which cluster
                method = "silhouette", max_cluster= 24, cluster_method = "kmeans", main= "Kmeans for 24 clusters in cluster 1")

plotoptimalcluster1 # optimal 1 break up cluster 1 in 9 clusters

GSEA.Object3breakup2 <- BreakUpCluster(GSEA.Object3breakup, 
                                      breakup.cluster = 1, # which cluster
                                      sub.cluster = 9, # in how many cluster split up
                                      uniquePathways = TRUE) # conside unique pathways

plotuniquebreakup2 <- PlotGeneSets(GSEA.Object3breakup2, 
                                   uniquePathways = TRUE, 
                                   wordcloud = FALSE, # wordcloud only supported for GO terms
                                   doORA = T) # do ora per cluster

plotuniquebreakup2

```

```{r breakup cluster number 1}
# let's say we are interested in exploring cluster 2 in plotunique. Lets break up this cluste for further analysis 

plotoptimalcluster2 <- OptimalGeneSets(Object = GSEA.Object3, 
                uniquePathway = TRUE, # consider all the pathways (also repeated) or the unique pathways
                cluster = 2, # which cluster
                method = "silhouette", max_cluster= 24, cluster_method = "kmeans", main= "Kmeans for 24 clusters in cluster 1")

plotoptimalcluster2 # optimal 2 break up cluster 2 in 2 clusters

GSEA.Object3breakup <- BreakUpCluster(GSEA.Object3, 
                                      breakup.cluster = 2, # which cluster
                                      sub.cluster = 2, # in how many cluster split up
                                      uniquePathways = TRUE) # conside unique pathways

plotuniquebreakup <- PlotGeneSets(GSEA.Object3breakup, 
                                  uniquePathways = TRUE, 
                                  wordcloud = FALSE, # wordcloud only supported for GO terms
                                  doORA = T) # do ora per cluster

plotuniquebreakup
```
